cpp/src/rolling/rolling.cu

```diff
@@ -623,5 +625,216 @@ std::unique_ptr<column> rolling_window(column_view const& input,
+          --preceding_i;
+        }
+
+        return idx - preceding_i + 1; // Preceding must account for current row.
+      };
+  
+    auto following_calculator = 
+      [
+        d_group_offsets = group_offsets.data().get(),
+        d_group_labels  = group_labels.data().get(),
+        d_timestamps    = timestamp_column.data<TimestampImpl_t>(),
+        following_window_in_days,
+        mult_factor
+      ] __device__ (size_type idx) {
+        auto group_label = d_group_labels[idx];
+        auto group_end = d_group_offsets[group_label+1]; // Cannot fall off the end, since offsets is capped with `input.size()`.
+        auto upper_bound = d_timestamps[idx] + following_window_in_days*mult_factor;
+
+        auto following_i{idx}; // Start at current row.
+        // If following value is part of current group, and does not exceed upper_bound, extend window forwards.
+        while ((following_i+1) < group_end && d_timestamps[following_i+1] <= upper_bound) {
+          ++following_i;
+        }
+
+        return following_i - idx;
+      };
+
+    return cudf::experimental::detail::rolling_window(
+      input,
+      thrust::make_transform_iterator(thrust::make_counting_iterator<size_type>(0), preceding_calculator),
+      thrust::make_transform_iterator(thrust::make_counting_iterator<size_type>(0), following_calculator),
+      min_periods, aggr, mr
+    );
+  }
+
+} // namespace detail;
+
+std::unique_ptr<column> grouped_time_range_rolling_window(table_view const& grouping_keys,
+                                                          column_view const& timestamp_column,
+                                                          column_view const& input,
+                                                          size_type preceding_window_in_days,
+                                                          size_type following_window_in_days,
+                                                          size_type min_periods,
+                                                          std::unique_ptr<aggregation> const& aggr,
+                                                          rmm::mr::device_memory_resource* mr ) {
+
+  if (input.size() == 0) return empty_like(input);
+
+  CUDF_EXPECTS((grouping_keys.num_columns() > 0), "Expected at least one grouping key.");
+
+  CUDF_EXPECTS((grouping_keys.num_rows() == input.size()), 
+               "Size mismatch between grouping_keys and input vector.");
+
+  CUDF_EXPECTS((min_periods > 0), "min_periods must be positive");
+
+  using sort_groupby_helper = cudf::experimental::groupby::detail::sort::sort_groupby_helper;
+  sort_groupby_helper helper{grouping_keys, cudf::include_nulls::NO, cudf::sorted::YES};
+
+  auto group_offsets{helper.group_offsets()};
+  group_offsets.push_back(input.size()); // Cap the end with the number of inputs.
+  auto const& group_labels{helper.group_labels()};
+
+  // Assumes that `group_offsets` starts with `0`, ends with `input.size`
+#ifndef NDEBUG
```

Comment:
Should have same expectations in release and debug.

Reply:
Consider `assert`.

Reply:
Yep, just as the previous `CUDF_EXPECTS`. I've changed this to an `assert()`. Will push a new revision.